core/shared/src/main/scala/zio/concurrent/ConcurrentSet.scala

```diff
@@ -11,15 +12,47 @@ final class ConcurrentSet[A] private (private val underlying: ConcurrentHashMap.
     UIO(underlying.add(x))
 
   def addAll(xs: Iterable[A]): UIO[Boolean] =
-    UIO {
-      var added = false
-      xs.foreach(x => added = underlying.add(x))
-      added
-    }
+    UIO(underlying.addAll(xs.asJavaCollection))
+
+  def collectFirst[B](pf: PartialFunction[A, B]): UIO[Option[B]] =
+    UIO(underlying.asScala.toSet.collectFirst(pf))
+
+  def exists(p: A => Boolean): UIO[Boolean] =
+    UIO(underlying.asScala.toSet.exists(p))
```

Comment:
To regain atomicity, I don't see another solution than to pay the cost of `forEach`. It would be possible to shortcut via `stream` I suppose but very unlikely this would be atomic?

Reply:
Ah yes! `forEach` would work better here.